becquerel/core/spectrum.py

```diff
@@ -806,6 +807,57 @@ def combine_bins(self, f):
         obj = Spectrum(**kwargs)
         return obj
 
+    def rebin(self, out_edges, method="interpolation", slopes=None,
+              zero_pad_warnings=True):
+        """
+        Spectra rebinning via deterministic or stochastic methods.
+
+        Args:
+            out_edges (np.ndarray): an array of the output bin edges
+                [num_channels_out]
+            method (str): rebinning method
+                "interpolation"
+                    Deterministic interpolation
+                "listmode"
+                    Stochastic rebinning via conversion to listmode of energies
+            slopes (np.ndarray|None): (optional) an array of input bin slopes
+                for quadratic interpolation
+                (only applies for "interpolation" method)
+                [len(spectrum) + 1]
+        zero_pad_warnings (boolean): warn when edge overlap results in
+            appending empty bins
+
+        Raises:
+            SpectrumError: for bad input arguments
+
+        Returns:
+            A new Spectrum object with the rebinned data.
+        """
+        if self.bin_edges_kev is None:
+            raise SpectrumError('Cannot rebin spectrum without energy '
+                                'calibration')
+        in_spec = self.counts_vals
+        if method.lower() == 'listmode':
+            if (self._counts is None) and (self.livetime is not None):
+                warnings.warn(
+                    'Rebinning by listmode method without explicit counts ' +
+                    'provided in Spectrum object',
+                    SpectrumWarning)
+            # NOTE: @jccurtis asks: why are the counts not integers already?
+            in_spec = in_spec.astype(np.int64)
```

Comment:
I added this intentionally because a warning is raised since the `Spectrum.count_vals` are `float` values. My note was a general question about why `Spectrum.count_vals` are `float`'s. You do bring up the obvious reason why this would be the case: `interpolation` rebinning ...

So we need to decide if `listmode` rebinning with `float` spectrum input raises a warning ...

Reply:
Okay one idea is to check the decimal residual on the float inputs:

```python
if np.any(np.abs(counts - np.fix(counts)) > 1e-6):
    raise RebinError()
```

along the same lines, if all values are less than one, raise an error:

```python
if np.all(counts < 1):
    raise RebinError()
```

Reply:
Isn't that what the check on line 399 in rebin.py is for?

I kind of like your idea about checking if the float values could originate from a integer array. However, it probably would make more sense in rebin.py or is this something that is specific to Spectrum somehow? Also I'm not sure how numpy handles float -> integer conversion, but it probably would be safer if we do np.fix(counts+0.5). Also could we use a np.allclose(counts - np.fix(counts+0.5)) or is there something fishy about that?